src/ATokenNaive.sol

```diff
@@ -0,0 +1,357 @@
+  IFractionalGovernor public immutable governor;
+
+  /// @dev Constructor.
+  /// @param _pool The address of the Pool contract
+  /// @param _governor The address of the flex-voting-compatable governance contract.
+  constructor(IPool _pool, address _governor) AToken(_pool) {
+    governor = IFractionalGovernor(_governor);
+  }
+
+  // TODO Is there a better way to do this? It cannot be done in the constructor
+  // because the AToken is just used a proxy -- it won't share an address with
+  // the implementation (i.e. this code).
+  function selfDelegate() public {
+    IVotingToken(governor.token()).delegate(address(this));
+  }
+
+  /// @notice Method which returns the deadline (as a block number) by which
+  /// depositors must express their voting preferences to this Pool contract. It
+  /// will always be before the Governor's corresponding proposal deadline.
+  /// @param proposalId The ID of the proposal in question.
+  function internalVotingPeriodEnd(uint256 proposalId)
+    public
+    view
+    returns (uint256 _lastVotingBlock)
+  {
+    _lastVotingBlock = governor.proposalDeadline(proposalId) - CAST_VOTE_WINDOW;
+  }
+
+  /// TODO how to handle onBehalfOf?
+  /// TODO should this revert if the vote has been cast?
+  /// @notice Allow a depositor to express their voting preference for a given
+  /// proposal. Their preference is recorded internally but not moved to the
+  /// Governor until `castVote` is called. We deliberately do NOT revert if the
+  /// internalVotingPeriodEnd has passed.
+  /// @param proposalId The proposalId in the associated Governor
+  /// @param support The depositor's vote preferences in accordance with the `VoteType` enum.
+  function expressVote(uint256 proposalId, uint8 support) external {
+    uint256 weight = getPastDeposits(msg.sender, governor.proposalSnapshot(proposalId));
+    if (weight == 0) revert("no weight");
+
+    if (_proposalVotersHasVoted[proposalId][msg.sender]) revert("already voted");
+    _proposalVotersHasVoted[proposalId][msg.sender] = true;
+
+    if (support == uint8(VoteType.Against)) {
+      proposalVotes[proposalId].againstVotes += SafeCast.toUint128(weight);
+    } else if (support == uint8(VoteType.For)) {
+      proposalVotes[proposalId].forVotes += SafeCast.toUint128(weight);
+    } else if (support == uint8(VoteType.Abstain)) {
+      proposalVotes[proposalId].abstainVotes += SafeCast.toUint128(weight);
+    } else {
+      revert("invalid support value, must be included in VoteType enum");
+    }
+  }
+
+  /// @notice Causes this contract to cast a vote to the Governor for all the
+  /// tokens it currently holds.  Uses the sum of all depositor voting
+  /// expressions to decide how to split its voting weight. Can be called by
+  /// anyone, but _must_ be called within `CAST_VOTE_WINDOW` blocks before the
+  /// proposal deadline.
+  /// @param proposalId The ID of the proposal which the Pool will now vote on.
+  function castVote(uint256 proposalId) external {
```

Comment:
I don't see any checks to prevent this from being called twice while within the window. Does the governor revert in that case, or does it just re-record the votes as a no-op?

Reply:
Yeah, the [governor will revert](https://github.com/ScopeLift/flexible-voting/blob/806eb9dd1d59af93053f7aeb6606534671bb02b1/src/GovernorCountingFractional.sol#L101), that's why we don't bother checking. I can add a note about this.

Reply:
Done in c324e66

Reply:
Ah, so if it reverts on subsequent votes, then I think we definitely need this check to prevent casting the votes until some internalVotingPeriodStart block/timestamp is reached? https://github.com/ScopeLift/flexible-voting/pull/11#discussion_r1020549163

Reply:
Sorry, I'm confused. We have that check right [here](https://github.com/ScopeLift/flexible-voting/pull/11/files#diff-5a237783c25a06fff902b80c7117e7e15796237c85e0a05dadc7b888afdcd8ecR135-R138)

Reply:
Oops you're right 😅